metaflow/plugins/kfp/kfp.py

```diff
@@ -168,123 +193,65 @@ def _get_resource_requirements(node):
                 value = f"{value}M"
             return value
 
-        resource_requirement = dict()
+        resource_requirements = {}
         for deco in node.decorators:
             if isinstance(deco, ResourcesDecorator):
                 for attr_key, attr_value in deco.attributes.items():
                     if attr_value is not None:
-                        resource_requirement[attr_key] = to_k8s_resource_format(
+                        resource_requirements[attr_key] = to_k8s_resource_format(
                             attr_key, attr_value
                         )
 
-        return resource_requirement
+        return resource_requirements
 
-    def create_kfp_components_from_graph(self):
+    def create_kfp_components_from_graph(self) -> Dict[str, KfpComponent]:
         """
-        Create a map of steps to their corresponding KfpComponent. The KfpComponent defines the component
-        attributes and step command to be used to run that particular step with placeholders for the `run_id`.
-
-        # Note:
-        # Level-order traversal is adopted to keep the task-ids in line with what happens during a local metaflow execution.
-        # It is not entirely necessary to keep this order of task-ids if we are able to point to the correct input-paths for
-        # each step. But, using this ordering does keep the organization of data in the datastore more understandable and
-        # natural (i.e., `start` gets a task id of 1, next step gets a task id of 2 and so on with 'end' step having the
-        # highest task id. So the paths in the datastore look like: {run-id}/start/1, {run-id}/next-step/2, and so on)
+        Returns a map of steps to their corresponding KfpComponent.
+        The KfpComponent defines the component attributes
+        and step command to be used to run that particular step.
         """
 
-        def build_kfp_component(node, step_name, task_id, input_paths):
+        def build_kfp_component(node: DAGNode, task_id: int):
             """
             Returns the KfpComponent for each step.
-
-            This method returns a string with placeholders for `run_id` and
-            `task_id` which get populated using the provided config and the kfp
-            run ID respectively.  The rest of the command string is populated
-            using the passed arguments which are known before the run starts.
-
-            An example constructed command template (to run a step named `hello`):
-            "python downloaded_flow.py --datastore s3 --datastore-root {ds_root} " \
-                             "step hello --run-id {run_id} --task-id 2 " \
-                             "--input-paths {run_id}/start/1"
             """
 
-            # TODO: @schedule, @environment, @resources, @timeout, @catch, etc.
-            # Resolve retry strategy.
+            # TODO: @schedule, @environment, @timeout, @catch, etc.
+            # TODO: @retry
```

Comment:
Hmm doesn't this code implement the retry mechanism? Do we still need the `TODO`?

Reply:
Not yet, hasn't been tested either...